library/alloc/src/str.rs

```diff
@@ -619,3 +637,60 @@ impl str {
 pub unsafe fn from_boxed_utf8_unchecked(v: Box<[u8]>) -> Box<str> {
     unsafe { Box::from_raw(Box::into_raw(v) as *mut str) }
 }
+
+/// Converts the bytes while the bytes are still ascii.
+/// This is performed in chunks of `16*size_of::<usize>()` which in testing
+/// performed the fastest.
+#[inline]
+#[cfg(not(test))]
+#[cfg(not(no_global_oom_handling))]
+fn convert_while_ascii(b: &[u8], out: &mut Vec<u8>, convert: fn(&u8) -> u8) {
+    out.reserve(b.len());
+
+    const USIZE_SIZE: usize = mem::size_of::<usize>();
+    const MAGIC_UNROLL: usize = 16;
```

Comment:
This is a... really aggressive unrolling, I'm a bit skeptical it's worth it. Keep in mind that this code gets statically linked most of the time, so we should keep in mind the code size cost of these things. (Not to mention that this pessimizes smaller strings, which are much more common candidates for case changes IME)

I feel like 2 or 4 tops is in line with how much we tend to unroll in the stdlib.

Reply:
Yeah that's fair. I'll knock it down and benchmark again with my short string setup

Reply:
Times are still decent so I'll stick with 2
```
# small strings (all ascii)
lowercase_small/case_conv                                                                            
                        time:   [143.71 us 143.95 us 144.20 us] *
lowercase_small/std_lib time:   [603.21 us 604.88 us 606.69 us]

# long strings (all ascii)
lowercase/ascii         time:   [35.872 us 35.927 us 35.986 us] *
lowercase/ascii_std     time:   [526.90 us 527.60 us 528.51 us]
```